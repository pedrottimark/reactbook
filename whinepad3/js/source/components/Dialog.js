/* @flow */

import React, { Component } from 'react';

import Button from './Button';

type Props = {
  header: string,
  confirmLabel: string,
  modal: boolean,
  onAction: Function,
  hasCancel: ?boolean,
  children?: Array<any>,
};

class Dialog extends Component {

  props: Props;

  static defaultProps = {
    confirmLabel: 'ok',
    modal: false,
    onAction: (_) => {},
    hasCancel: true,
  };

  componentWillUnmount() {
    document.body.classList.remove('DialogModalOpen');
  }

  componentDidMount() {
    if (this.props.modal) {
      document.body.classList.add('DialogModalOpen');
    }
  }

  render() {
    const { children, confirmLabel, hasCancel, header, modal, onAction } = this.props;
    return (
      <div className={modal ? 'Dialog DialogModal' : 'Dialog'}>
        <div className={modal ? 'DialogModalWrap' : null}>
          <div className="DialogHeader">{header}</div>
          <div className="DialogBody">{children}</div>
          <div className="DialogFooter">
            {hasCancel
              ? <span
                  className="DialogDismiss"
                  onClick={onAction.bind(null, false)}>
                  Cancel
                </span>
              : null
            }
            <Button onClick={onAction.bind(null, hasCancel)}>
              {confirmLabel}
            </Button>
          </div>
        </div>
      </div>
    );
  }
}

export default Dialog
