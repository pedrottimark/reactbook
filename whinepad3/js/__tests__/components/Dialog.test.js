import React from 'react';
import ReactDOM from 'react-dom';
import TestUtils from 'react-addons-test-utils';
import renderer from 'react-test-renderer';

import Dialog from '../../source/components/Dialog';

describe('Dialog with hasCancel={true} or no hasCancel prop', () => {

  it('renders ok and cancel buttons', () => {
    const tree = renderer.create(
      <Dialog header="ok and Cancel" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders confirmLabel prop as text of the primary button', () => {
    const tree = renderer.create(
      <Dialog header="Add and Cancel" confirmLabel="Add" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('returns true or false when people click the primary or secondary button', () => {
    const callback = jest.fn();
    const component = TestUtils.renderIntoDocument(
      <Dialog header="ok and Cancel" onAction={callback} />
    );
    TestUtils.Simulate.click(
      TestUtils.findRenderedDOMComponentWithTag(component, 'button')
    );
    TestUtils.Simulate.click(
      TestUtils.findRenderedDOMComponentWithTag(component, 'span')
    );

    const calls = callback.mock.calls;
    expect(calls.length).toEqual(2);
    expect(calls[0][0]).toEqual(true);
    expect(calls[1][0]).toEqual(false);
  });
});

describe('Dialog with hasCancel={false} prop', () => {

  it('renders only one button', () => {
    const tree = renderer.create(
      <Dialog header="ok" hasCancel={false} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders confirmLabel prop as text of the button', () => {
    const tree = renderer.create(
      <Dialog header="Close" hasCancel={false} confirmLabel="Close" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('returns false when people click the button', () => {
    const callback = jest.fn();
    const component = TestUtils.renderIntoDocument(
      <Dialog header="Close" hasCancel={false} confirmLabel="Close" onAction={callback} />
    );
    TestUtils.Simulate.click(
      TestUtils.findRenderedDOMComponentWithTag(component, 'button')
    );

    const calls = callback.mock.calls;
    expect(calls.length).toEqual(1);
    expect(calls[0][0]).toEqual(false);
  });

});

describe('Dialog with modal={true} prop', () => {

  it('adds and removes class to body when it mounts and unmounts', () => {
    const dialog = TestUtils.renderIntoDocument(
      <Dialog header="Modal" modal={true}>Civilized dialog</Dialog>
    );
    expect(Array.from(document.body.classList)).toContain('DialogModalOpen');

    // removing the dialog
    ReactDOM.unmountComponentAtNode(ReactDOM.findDOMNode(dialog).parentNode);
    expect(Array.from(document.body.classList)).not.toContain('DialogModalOpen');
  });

});
