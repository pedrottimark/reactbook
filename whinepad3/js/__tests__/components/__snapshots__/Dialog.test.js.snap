exports[`Dialog with hasCancel={false} prop renders confirmLabel prop as text of the button 1`] = `
<div
  className="Dialog">
  <div
    className={null}>
    <div
      className="DialogHeader">
      Close
    </div>
    <div
      className="DialogBody" />
    <div
      className="DialogFooter">
      <button
        className="Button"
        onClick={[Function]}>
        Close
      </button>
    </div>
  </div>
</div>
`;

exports[`Dialog with hasCancel={false} prop renders only one button 1`] = `
<div
  className="Dialog">
  <div
    className={null}>
    <div
      className="DialogHeader">
      ok
    </div>
    <div
      className="DialogBody" />
    <div
      className="DialogFooter">
      <button
        className="Button"
        onClick={[Function]}>
        ok
      </button>
    </div>
  </div>
</div>
`;

exports[`Dialog with hasCancel={true} or no hasCancel prop renders confirmLabel prop as text of the primary button 1`] = `
<div
  className="Dialog">
  <div
    className={null}>
    <div
      className="DialogHeader">
      Add and Cancel
    </div>
    <div
      className="DialogBody" />
    <div
      className="DialogFooter">
      <span
        className="DialogDismiss"
        onClick={[Function]}>
        Cancel
      </span>
      <button
        className="Button"
        onClick={[Function]}>
        Add
      </button>
    </div>
  </div>
</div>
`;

exports[`Dialog with hasCancel={true} or no hasCancel prop renders ok and cancel buttons 1`] = `
<div
  className="Dialog">
  <div
    className={null}>
    <div
      className="DialogHeader">
      ok and Cancel
    </div>
    <div
      className="DialogBody" />
    <div
      className="DialogFooter">
      <span
        className="DialogDismiss"
        onClick={[Function]}>
        Cancel
      </span>
      <button
        className="Button"
        onClick={[Function]}>
        ok
      </button>
    </div>
  </div>
</div>
`;
